### PR TITLE
QUnit.stack

### PR DESCRIPTION
**Pull request by leobalter**
Ref #758

### CODE REVIEW

**Comment by leobalter** [comment]
Now this branch is integrated with the changes on #758 and it's finally passing over all the browserstack tests.

**Comment by leobalter** [comment]
browserstack tests are passing, but I'm getting a timeout on TravisCI.

As told by @jzaefferer in a private chat, it's good to have one other opinion here.

@scottgonzalez, would you review it, please?

File: test/main/stack.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+( function() {
+	var stack = QUnit.stack();
+
+	QUnit.module( "QUnit.stack" );
+
+	// Flag this test as skipped on browsers that doesn't support stack trace
+	QUnit[ stack ? "test" : "skip" ]( "returns the proper stack line", function( assert ) {
+		assert.ok( /\/test\/main\/stack\.js/.test( stack ) );
+	} );
+
```
Comment (scottgonzalez) [inline test/main/stack.js]: extraneous blank line

File: test/reporter-html/reporter-html.js [hunk 2]
```diff
@@ -86,10 +42,22 @@ QUnit.test( "running test name displayed", function( assert ) {
 });
 
 QUnit.module( "timing", {
-	beforeEach: function( assert ) {
+	getPreviousTest: function( assert ) {
+		return document.getElementById( "qunit-test-output-" + assert.test.testId  )
+				.previousSibling;
```
Comment (scottgonzalez) [inline test/reporter-html/reporter-html.js]: extra indentation

File: test/reporter-html/reporter-html.js [hunk 3]
```diff
@@ -100,27 +68,61 @@ QUnit.module( "timing", {
...
+			false,
+			"Don't add source information on non-supported environments"
+		);
+		return;
+	}
+
+	assert.ok( /(^| )qunit-source( |$)/.test( source.className ), "Source element exists" );
+	assert.equal( source.firstChild.innerHTML, "Source: " );
+
+	// test/reporter-html.js is a direct reference to this test file
```
Comment (scottgonzalez) [inline test/reporter-html/reporter-html.js]: The path is wrong in the comment.

File: test/reporter-html/reporter-html.js [hunk 4]
```diff
@@ -100,27 +68,61 @@ QUnit.module( "timing", {
...
-})();
+QUnit.module( "source" );
+
+QUnit.test( "setup", function( assert ) {
+	assert.expect( 0 );
+});
+
+QUnit.test( "logs location", function( assert ) {
+	var previous = document.getElementById( "qunit-test-output-" + assert.test.testId  )
+			.previousSibling;
```
Comment (scottgonzalez) [inline test/reporter-html/reporter-html.js]: extra indentation

File: test/reporter-html/reporter-html.js [hunk 5]
```diff
@@ -100,27 +68,61 @@ QUnit.module( "timing", {
...
+	var source = previous.lastChild;
+	var stack = QUnit.stack();
+
+	// Verify QUnit supported stack trace
+	if ( !stack ) {
+		assert.equal(
+			/(^| )qunit-source( |$)/.test( source.className ),
+			false,
+			"Don't add source information on non-supported environments"
```
Comment (scottgonzalez) [inline test/reporter-html/reporter-html.js]: non-supported -> unsupported

File: test/reporter-html/reporter-html.js (hunk 4 above)
Comment (leobalter) [inline test/reporter-html/reporter-html.js]: https://www.dropbox.com/s/mvei73ejuekjt4o/Captura%20de%20tela%202015-03-27%2014.07.43.png?dl=0

a line break after a var declaration doesn't require an extra indentation?

Comment (leobalter) [inline test/reporter-html/reporter-html.js]: I was wrong: http://contribute.jquery.org/style-guide/js/#multi-line-statements

**Comment by scottgonzalez** [comment]
I'm getting a test failure in local Chrome.

**Comment by leobalter** [comment]
Did you build it with grunt? It's passing here.

UA: 
- Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2272.104 Safari/537.36
- `Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:36.0) Gecko/20100101 Firefox/36.0`

**Comment by scottgonzalez** [comment]
> Did you build it with grunt?

yes

**Comment by scottgonzalez** [comment]
The failure was caused by running previously failed tests first. We should probably disable test reordering since a lot of these tests rely on a specific order.

**Comment by leobalter** [comment]
the logs and reporter-html tests rely on the order. I'm +1000 to disable the test reordering

**Comment by leobalter** [comment]
@scottgonzalez b0ad27f now test/reporter-html/index.html will avoid tests reordering.

File: reporter/html.js [hunk 6]
```diff
@@ -754,6 +754,20 @@ QUnit.testDone(function( details ) {
 		time.innerHTML = details.runtime + " ms";
 		testItem.insertBefore( time, assertList );
 	}
+
+	// Only add this information if stackTrace is supported
```
Comment (jzaefferer) [inline reporter/html.js]: Replace this comment with something explaining why this is here, or what it does, something like:

```js
// Show the source of the test when showing assertions
```

**Comment by jzaefferer** [comment]
The paragraph showing the test's source has too much margin, something like this looks much better: `margin: 7px 0px 4px;`

**Comment by jzaefferer** [comment]
We also need to write API docs for the new method, could you create a PR for that?

**Comment by leobalter** [comment]
> could you create a PR for that?

yes, sure

**Comment by leobalter** [comment]
The source comment commit should be squashed into 477e1f1

**Comment by jzaefferer** [comment]
This looks good. Apparently Travis didn't run the more recent commits, so can you run browserstack-runner manually to see if everything passes? When it does, let's merge.

**Comment by leobalter** [comment]
did another change: QUnit.stack default public offset is now 0, with tests on d73e81a.

**Comment by jzaefferer** [comment]
Looks good. Build failed, but all with timeouts on unrelated testsuites. Still have to address them, but this can be merged.

**Comment by leobalter** [comment]
It finally passed! Browserstack is also timing out a lot on my local env.

**Pull request closed without merging**